### ISSUE TITLE
Batching rules of operations related to BatchNorm

### ISSUE THREAD

**Issue by YiteWang** [facebookresearch/functorch#115]
Dear Facebook researchers,

We are trying to use `functorch` to compute per-sample-gradient for complex architectures which include Batchnorm layers and many other operations. However, currently there are several operations not supported by `functorch`. So we have to use other packages and replace Batchnorm with other normalization layers. However, we really want to use BatchNorm for best performance, thus I am raising an issue here hoping the following operations could be implemented.

`aten::thnn_conv2d_forward`, 
`aten::native_batch_norm`, 
`aten::native_batch_norm_backward`, 
`aten::cudnn_batch_norm`, 
`aten::slow_conv_dilated2d`, 
`aten::slow_conv_dilated2d_backward`.
`aten::_conv_depthwise2d_backward.output_mask`, 
`aten::_conv_depthwise2d`, 
`aten::avg_pool2d_backward`, 

We are not in a hurry and I totally understand that there are a lot of other operations need to be implemented, but it would be much appreciated if these operations can be prioritized.

Thank you so much!

**Comment by zou3519** [comment]
Yes, we can prioritize the above. The convolution ones and the pooling ones are definitely on our list.

batch_norm for per-sample-grad computation has some weird semantics. Do you expect it to normalize across the batch, or to normalize a single element in the batch?

For example, if we take the definition of "per-example-gradient" to be "forward and backward a single data point through the model", then it turns out that the batch norm is almost equivalent to an instance norm 
```py
model = some_model_with_batch_norm()
data = torch.randn(batch_size, 64)
targets = torch.randn(batch_size)
loss = loss_fn(model(data[0].unsqueeze(-1)), targets[0].unsqueeze(-1))
data_0_grad = torch.autograd.grad(loss, model.parameters())
```

**Comment by YiteWang** [comment]
Dear Facebook researchers,

I expect it to normalize across the batch, which means it should generate per-sample-gradient like the following:

```
model = some_model_with_batch_norm()
data = torch.randn(batch_size, 64)
output = model(data)
targets = torch.randn(batch_size)
loss_0 = some_loss(output[0],targets[0]) / batch_size
data_0_grad = torch.autograd.grad(loss_0, model.parameters())
```

Thank you so much!

**Comment by Chillee** [comment]
@YiteWang Out of curiosity, are you running on GPU or CPU? Or are you running on both?

**Comment by YiteWang** [comment]
> @YiteWang Out of curiosity, are you running on GPU or CPU? Or are you running on both?

I am running it on GPU.

**Comment by Chillee** [comment]
How did you get the operators that need batching rules? On CPU the operators dispatch to different operators from GPU. For example, if you only care about GPU batching rules, we don't need to care about the first 3 :)

**Comment by YiteWang** [comment]
> How did you get the operators that need batching rules? On CPU the operators dispatch to different operators from GPU. For example, if you only care about GPU batching rules, we don't need to care about the first 3 :)

Thank you for pointing out! For first three operators, I got those when I was testing part of the network using CPU.
And thus we do not need first three operators.

**Comment by zou3519** [comment]
> Dear Facebook researchers,
> 
> I expect it to normalize across the batch, which means it should generate per-sample-gradient like the following:
> 
> ```
> model = some_model_with_batch_norm()
> data = torch.randn(batch_size, 64)
> output = model(data)
> targets = torch.randn(batch_size)
> loss_0 = some_loss(output[0],targets[0]) / batch_size # The division may not be necessary
> data_0_grad = torch.autograd.grad(loss_0, model.parameters())
> ```
> 
> Thank you so much!

How important to you is using BatchNorm (or that this behavior happens?) On our side this is a little tricky to implement because vmap assumes there are no inter-batch operations (and the above behavior of BatchNorm requires inter-batch operations)

**Comment by YiteWang** [comment]
Dear Facebook researchers,

I totally agree that such inter-batch operation is very tricky and we believe such inter-batch behavior is very important. We could only find two packages for efficient per-sample-gradient computations, which include Opacus and functorch. And actually we are using Opacus now to compute per-sample-gradient with batchnorms replaced by groupnorm/instancenorm. However, I believe Opacus will never allow inter-batch behaviors since it is invented for differential privacy(DP). Thus I am raising an issue here hoping such behavior could be implemented.

However, we would still be very grateful even if functorch only supports normalizing a single element in the batch.

Thank you so much!
